consensus/parlia/parlia.go

```diff
@@ -1812,14 +1812,43 @@ func backOffTime(snap *Snapshot, val common.Address) uint64 {
 			// The backOffTime does not matter when a validator is not authorized.
 			return 0
 		}
+
 		s := rand.NewSource(int64(snap.Number))
 		r := rand.New(s)
 		n := len(snap.Validators)
-		backOffSteps := make([]uint64, 0, n)
-		for idx := uint64(0); idx < uint64(n); idx++ {
-			backOffSteps = append(backOffSteps, idx)
+		if !p.chainConfig.IsBoneh(header.Number) {
+			backOffSteps := make([]uint64, 0, n)
+			for i := uint64(0); i < uint64(n); i++ {
+				backOffSteps = append(backOffSteps, i)
+			}
+			r.Shuffle(n, func(i, j int) {
+				backOffSteps[i], backOffSteps[j] = backOffSteps[j], backOffSteps[i]
+			})
+			delay := initialBackOffTime + backOffSteps[idx]*wiggleTime
+			return delay
+		}
+
+		// Exclude the recently signed validators first, and then compute the backOffTime.
+		recentVals := make(map[common.Address]bool, len(snap.Recents))
+		for _, recent := range snap.Recents {
+			if val == recent {
+				// The backOffTime does not matter when a validator has signed recently.
+				return 0
+			}
+			recentVals[recent] = true
+		}
+		backOffSteps := make([]uint64, 0, n-len(snap.Recents))
+		validators := snap.validators()
+		for i := uint64(0); i < uint64(n); i++ {
+			if isRecent, ok := recentVals[validators[i]]; ok && isRecent {
+				if i < uint64(idx) {
+					idx--
```

Comment:
Too many duplicated code here. The only difference is that, the validator set that been shuffled is different, please reduce the code.